common/src/main/java/com/microsoft/identity/common/internal/logging/Logger.java

```diff
@@ -116,71 +119,395 @@ public void setExternalLogger(final ILoggerCallback externalLogger) {
+     * @param message       The message to log.
      */
-    public static void verbose(final String tag, final String correlationID, final String message) {
-        getInstance().log(tag, LogLevel.VERBOSE, correlationID, message, null, false);
+    public static void verbose(final String tag,
+                               @Nullable final String correlationID,
+                               @Nullable final String message) {
+        getInstance().log(
+                tag,
+                LogLevel.VERBOSE,
+                correlationID,
+                message,
+                null,
+                false
+        );
     }
 
     /**
      * Send a {@link LogLevel#VERBOSE} log message with PII.
+     *
+     * @param tag     Used to identify the source of a log message. It usually identifies the class
+     *                or activity where the log call occurs.
+     * @param message The message to log.
+     */
+    public static void verbosePII(final String tag, @Nullable final String message) {
+        getInstance().log(
+                tag,
+                LogLevel.VERBOSE,
+                DiagnosticContext.getRequestContext().toJsonString(),
+                message,
+                null,
+                true
+        );
+    }
+
+    /**
+     * Send a {@link LogLevel#VERBOSE} log message with PII.
+     *
+     * @param tag           Used to identify the source of a log message. It usually identifies the
+     *                      class or activity where the log call occurs.
+     * @param correlationID Unique identifier for a request or flow used to trace program execution.
+     * @param message       The message to log.
      */
-    public static void verbosePII(final String tag, final String correlationID, final String message) {
-        getInstance().log(tag, LogLevel.VERBOSE, correlationID, message, null, true);
+    public static void verbosePII(final String tag,
+                                  @Nullable final String correlationID,
+                                  @Nullable final String message) {
+        getInstance().log(
+                tag,
+                LogLevel.VERBOSE,
+                correlationID,
+                message,
+                null,
+                true
+        );
     }
 
     /**
-     * TODO. Need to discuss on how to keep the correlationID. CorrelationID should be per request => need to sync with Telemetry implementation
+     * Log a method entry, with optional parameters. This method only writes output if the
+     * {@link LogLevel} is {@link LogLevel#VERBOSE} and PII logging is enabled.
      *
+     * @param tag    Used to identify the source of a log message. It usually identifies the class
+     *               or activity where the log call occurs.
+     * @param method The method which is being entered.
+     * @param args   The arguments to the method being entered.
+     */
+    public static void entering(final String tag,
```

Comment:
This is quite a big change on Logging strategy. I'm not convinced why these two methods are needed.

Reply:
These calls will only be expensive if PII && verbose logging are enabled; if those conditions aren't satisfied [nothing happens](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/logging/Logger.java#L179-L181) (inexpensive).

These addition of `entering()` and `exiting()` are actually unrelated to tracking the `correlationId` -- I'll update the title to reflect this.

I appreciate this represents a change of pace for us (as ADAL), but logging method entry/exit at the finest log-granularity isn't a faux-pas; there are a number of frameworks which provide this functionality out-of-the-box including [`java.util.logging.Logger`](https://developer.android.com/reference/java/util/logging/Logger.html), [Hugo](https://github.com/JakeWharton/hugo), and [Log4J](https://logging.apache.org/log4j/2.x/manual/flowtracing.html).

Regarding how this gets integrated, whomever calls into the common cache is expected to set the `correlationId` of their request on their thread. As long as execution stays on that thread, the value persists but is thread-private (so other callers can set their own). If execution changes thread, that value must be passed to the new thread ([do-able](https://stackoverflow.com/questions/7259906/propagating-threadlocal-to-a-new-thread-fetched-from-a-executorservice?noredirect=1&lq=1)).

When you get a `correlationId` you set it on a `RequestContext` and you give that to the `DiagnosticContext`, the rest is "free"

```java
final IRequestContext rc= new RequestContext();
rc.put("correlation_id", generateCorrelationId()); // it's a Map, so you can actually jam anything you want in here and it will show up in the logs as JSON

DiagnosticContext.setRequestContext(rc);
```

I don't expect every method of ADAL or MSAL to hook into this immediately --  for the purposes of this PR, I've set it up for `MsalOauth2TokenCache` -- you see the `entering()` and `exiting()` call at the beginning/end of each method

Reply:
I agree with Brian that this doesn't have to be all in one go.  I also agree with leveraging DiagnosticContext for passing things like correlation id.  I'd consider going further with that once we've agreed on what the threading model will look like and which component will manage our thread pools (silent and interactive)  I'm not concerned about performance since these methods will be a no-op if verbose logging (tracing) is not enabled.

Reply:
Thanks a lot for the detailed explaination, Brian. 😄  Yes, agree, as well as the leverage of DiagnosticContext. 

And here are two concerns/confusion for me.
1. `entering()` and `existing()` are calling `verbosePII()`. It means if the PII logging is disabled, all these info will not be logged. Is it by design that `entering()` and `existing()` are only used for verbose PII enabled logging?

2. What will happen if `entering()` gets called again insides other methods before calling `existing()`?